sdks/java/core/src/main/java/org/apache/beam/sdk/util/Timer.java

```diff
@@ -60,8 +60,35 @@
   void setForNowPlus(Duration durationFromNow);
 
   /**
+   * return a {@link AlignTimer} that aligns a timestamp to the next boundary of {@code period}.
+   */
+  AlignTimer align(Duration period);
+
+  /**
    * Unsets this timer. It is permitted to {@code cancel()} whether or not the timer was actually
    * set.
    */
   void cancel();
+
+  /**
+   * Sets or resets the time aligned to the smallest multiple of {@code period} since the
+   * {@code offset} greater than the current time in the timer's {@link TimeDomain} at
+   * which this it should fire. If the timer was already set, resets it to the new requested time.
+   *
+   * <p>For {@link TimeDomain#EVENT_TIME}, to prevent more than window GC Time, it provide a simple
+   * utility to take min(time to set, GC Time of window).
+   */
+  interface AlignTimer {
+
+    /**
+     * Sets or resets the time relative to the current time in the timer's {@link TimeDomain}.
+     */
+    void setForNow();
```

Comment:
To reduce the number of interfaces, I think we can combine `Timer` and `AlignTimer` to make this:

```java
interface Timer {
  void set(Instant absoluteTime);
  void cancel();

  // Sets the timer relative to the current time, according to any offset and alignment specified
  // using offset() and align()
  void setRelative();
  Timer offset(Duration offset);
  Timer align(Duration period);
}
```